bundles/org.openhab.binding.icalendar/src/main/java/org/openhab/binding/icalendar/internal/logic/BiweeklyPresentableCalendar.java

```diff
@@ -0,0 +1,348 @@
+            if (duration == null) {
+                continue;
+            }
+            // iterate over all begin dates
+            DateIterator begDates = getRecurredEventDateIterator(event);
+            while (begDates.hasNext()) {
+                Instant begInst = begDates.next().toInstant();
+                Instant endInst = begInst.plus(duration);
+                if (endInst.isBefore(frameBegin)) {
+                    continue;
+                } else if (endInst.isAfter(frameEnd)) {
+                    break;
+                }
+                // fall through => means we are within the time frame
+                if (eventList == null) {
+                    eventList = new ArrayList<Event>();
+                }
+                eventList.add(new VEventWPeriod(event, begInst, endInst).toEvent());
+                break;
+            }
+        }
+        return eventList;
+    }
+
+    @Override
+    public @Nullable Event getNextEvent(Instant instant) {
+        Collection<VEventWPeriod> candidates = new ArrayList<VEventWPeriod>();
+        Collection<VEvent> negativeEvents = new ArrayList<VEvent>();
+        Collection<VEvent> positiveEvents = new ArrayList<VEvent>();
+        classifyEvents(positiveEvents, negativeEvents);
+        for (VEvent currentEvent : positiveEvents) {
+            DateIterator startDates = this.getRecurredEventDateIterator(currentEvent);
+            Duration duration = getEventLength(currentEvent);
+            if (duration == null) {
+                continue;
+            }
+            startDates.advanceTo(Date.from(instant));
+            while (startDates.hasNext()) {
+                Instant startInstant = startDates.next().toInstant();
+                if (startInstant.isAfter(instant)) {
+                    @Nullable
+                    Uid currentEventUid = currentEvent.getUid();
+                    if (currentEventUid == null || !isCounteredBy(startInstant, currentEventUid, negativeEvents)) {
+                        candidates.add(new VEventWPeriod(currentEvent, startInstant, startInstant.plus(duration)));
+                        break;
+                    }
+                }
+            }
+        }
+        VEventWPeriod earliestNextEvent = null;
+        for (VEventWPeriod positiveCandidate : candidates) {
+            if (earliestNextEvent == null || earliestNextEvent.start.isAfter(positiveCandidate.start)) {
+                earliestNextEvent = positiveCandidate;
+            }
+        }
+
+        if (earliestNextEvent == null) {
+            return null;
+        }
+        return earliestNextEvent.toEvent();
+    }
+
+    @Override
+    public boolean isEventPresent(Instant instant) {
+        return (this.getCurrentComponentWPeriod(instant) != null);
+    }
+
+    /**
+     * A Class describing an event together with a start and end instant.
+     *
+     * @author Michael Wodniok - Initial contribution.
+     */
+    private static class VEventWPeriod {
```

Comment:
nested classes should be defined at the end of the file

Reply:
OK